packages/tic-tac-toe/contracts/tests/types/app-data.ts

```diff
@@ -0,0 +1,142 @@
+// import {Weapon} from './weapons';
+import {defaultAbiCoder, bigNumberify} from 'ethers/utils';
+import {unreachable} from '../utils/unreachable';
+
+export enum PositionType {
+  Start, // 0
+  XPlaying, // 1
+  OPlaying, // 2
+  Draw, // 4
+  Victory // 5
+}
+
+export interface TTTData {
+  positionType: PositionType;
+  stake: string; // uint256
+  Xs: number; // uint16
+  Os: number; // uint16
+}
+export interface Start {
+  type: 'start';
+}
+export interface XPlaying {
+  type: 'xPlaying';
+  stake: string;
+  Xs: number;
+}
+
+export interface OPlaying {
+  type: 'oPlaying';
+  stake: string;
+  Os: number;
+}
+
+export interface Draw {
+  type: 'draw';
+  stake: string;
+  Xs: number;
+  Os: number;
+}
+
+export interface Victory {
+  type: 'victory';
+  stake: string;
+  Xs: number;
+  Os: number;
+}
+
+export type AppData = Start | XPlaying | OPlaying | Draw | Victory;
+
+function toTTTData(appData: AppData): TTTData {
+  let positionType;
+  switch (appData.type) {
+    case 'start':
+      positionType = PositionType.Start;
+      break;
+    case 'xPlaying':
+      positionType = PositionType.XPlaying;
+      break;
+    case 'oPlaying':
+      positionType = PositionType.OPlaying;
+      break;
+    case 'draw':
+      positionType = PositionType.Draw;
+      break;
+    case 'victory':
+      positionType = PositionType.Victory;
+      break;
+  }
+  const defaults: TTTData = {
+    positionType,
+    stake: bigNumberify(0).toString(),
+    Xs: 0,
+    Os: 0
+  };
+
+  return {...defaults, ...appData};
+}
+
+export function encodeTTTData(tttData: TTTData): string {
+  return defaultAbiCoder.encode(
+    ['tuple(uint8 positionType, uint256 stake, uint16 Xs, uint16 Os)'],
+    [tttData]
+  );
+}
+
+export function encodeAppData(appData: AppData): string {
+  return encodeTTTData(toTTTData(appData));
+}
+
+export function decodeAppData(appDataBytes: string): AppData {
```

Comment:
Be good to write some tests for these `encoders`/`decoders`.

Reply:
Could we wait with this? Not really sure how to test this and I want to get this merged and start going on the client :)

Reply:
Yes we can wait with this. Over in `rps` I think we do something along the line of `expect(x).toEqual(decode(encode(x)))`.